TEMPO NO2 reader

blychs opened the pull request:
This is a tempo reader based on the TROPOMI NO2 reader, set up for compatibility with MELODIES-MONET. It has been tested with locally downloaded data, using MONET plotting capabilities. Pairing (at the MONET or MELODIES-MONET levels) has not been programmed in yet. A test case for this, with automatic downloads, would require an earthdata user --which would also require an additional dependency (namely, earthdata). If that is acceptable, I could set it up.
Pablo

blychs committed:
TEMPO NO2 reader

zmoon commented:
@blychs , alternatively, if you have an example file, we could place it here: https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_observation_data/satellite/

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,146 @@
+def open_dataset(fnames, variable_dict, debug=False):
+    if debug:
+        logging_level = logging.DEBUG
+        logging.basicConfig(strea=sys.stdout, level=logging_level)
```
```suggestion
        logging.basicConfig(stream=sys.stdout, level=logging_level)
```

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,146 @@
+        values_var = dso.groups["product"][varname]
+        values = values_var[:].squeeze()
+        fv = values_var.getncattr("_FillValue")
+        if not np.isfinite(fv):
```
What is the rationale behind this condition? `_FillValue` could be set to anything.

zmoon replied:
For example, the default nc fill value for double, even though it's large, is still finite, so the replacement by NaN will be skipped with this current `if not np.isfinite(fv):` guard.
```python
>>> np.isfinite(netCDF4.default_fillvals["f8"])
True
```

blychs replied:
Done.

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,146 @@
+        time_var[:].squeeze(),
+        {"long_name": time_var.long_name, "units": time_var.units},
+    )
+    ds["lon"] = ds["lon"].fillna(-999.99)
+    ds["lat"] = ds["lat"].fillna(-999.99)
```
Why not leave them as NaN?

blychs replied:
Because matplotlib.pcolormesh breaks if there are NaN values in the coordinates (the _values_ can be NaN, but not the coordinates). However, the lat/lon that are NaN in the original arrays are because the satellite is pointing at space, and therefore lat and lon are not defined.


 Another solution might be using pcolor, which is the oficially suggested solution (https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.pcolormesh.html, look at the section on masked arrays), but that would break away from the behavior of draw_map() in MONET. 
The workaround of turning them into -999.99 was taken from Amy K. Huff (https://gist.github.com/akhuff157/8411788ed06d459972f9b15203c8d793), but if you have a better solution, I am happy to change this.

EDIT: Clarify the pcolor suggested solution
EDIT2: correct typo

zmoon replied:
Since this issue is more to do with plotting, I would prefer this not be done in the reader. Perhaps it could be done in MM before pcolormesh is used. Or, we could do it here but make it an option that is off by default.

blychs replied:
Let's do it at the MM level.
I removed those lines in [fb4f571](https://github.com/noaa-oar-arl/monetio/pull/188/commits/fb4f571e86dbd3348b61eeb6904435a335c6d5d0)

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,146 @@
+
+def apply_quality_flag(ds):
+    """Mask variables in place based on quality flag
+    Parameters
+    ----------
+    ds: xr.Dataset
```
```suggestion

    Parameters
    ----------
    ds : xr.Dataset
```

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,146 @@
+
+def _open_one_dataset(fname, variable_dict):
+    """Read locally stored INTELSAT TEMPO NO2 level 2
+    Parameters
+    ----------
+    fname: string
+        fname is local path to netcdf4 (HDF5) file
+
+    variable_dict: dict
+    Returns
+    -------
+    ds: xarray dataset
```
```suggestion
    Parameters
    ----------
    fname : str
        Local path to netCDF4 (HDF5) file.
    variable_dict : dict

    Returns
    -------
    ds : xr.Dataset
```
more info: https://numpydoc.readthedocs.io/en/latest/format.html

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs commented:
> @blychs , alternatively, if you have an example file, we could place it here: https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_observation_data/satellite/

I do have files for this. Let me know how to give them to you so that they can be uploaded (I would just need to get rid of unnecessary variables so that we can have a full scan).

zmoon commented:
> I do have files for this. Let me know how to give them to you so that they can be uploaded (I would just need to get rid of unnecessary variables so that we can have a full scan).

Sounds good. I'll send an email.

blychs commented:
I found a bug in my code, which is with the application of the quality flags, maximum and minimum. i. e., I'm asking it to compare the values with the max, min or (for the quality_flag) qa_treshold. However, this breaks if the value is NaN, due to the way the comparison is made. It should be an easy fix, which I'll hopefully update today.

blychs committed:
Fix bug with flags and add options for vars

	Previous version failed when nan was compared with flags/max/min
	Added options for more variables, including pressure calculation

blychs commented:
Those bugs should be fixed now. I also added options for the rest of the possible variables a user might want, including everything needed to apply air mass factors and scattering weights and a pressure calculations. This means that the changes are quite large, and might require careful checking. I tried my code with all variables and it seems to be working fine, and the pre-commit hooks are passing. I don't really know how to run the linter though, should I just install pylint and give it a go?

zmoon commented:
>  the pre-commit hooks are passing. I don't really know how to run the linter though

Linting is included in the pre-commit hooks. But if you do want to run it (flake8) individually sometime:
```
pre-commit run flake8 --all-files
```
or
```
pre-commit run flake8 --file some/path.py
```

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,233 @@
+
+        if "minimum" in variable_dict[varname]:
+            minimum = variable_dict[varname]["minimum"]
+            values[:] = np.where(values[:] >= minimum, values[:], np.nan)
```
With this change, if `minimum` is NaN, `values` gets fully filled with NaN (whereas before it would do nothing). Is that really the desired behavior? Why would one set minimum/maximum to NaN anyway?

blychs replied:
The problem before was when minimum was a number (either int or float) and some values were NaN: the code would just break saying that it could not do the comparison, at least with Python 3.11. 

I hadn't realized the problem of minimum being NaN (no one should set it to NaN, but I guess it still needs to be addressed), and was addressing the case of:
a) minimum not being there (in which case the np.where() function is not applied),
b) minimum being a number, in which case this works.

I'll add a check for minimum being NaN to avoid this.

zmoon replied:
> The problem before was when minimum was a number (either int or float) and some values were NaN: the code would just break saying that it could not do the comparison, at least with Python 3.11.

I tried to replicate this with Python 3.11 / NumPy 2.1.0 and was not able to. Is this not the correct situation?
```python
>>> a = np.r_[1.0, 2.0, np.nan, 4.0]
>>> a[a < 2] = np.nan
>>> a
array([nan,  2., nan,  4.])
```

blychs replied:
OK, this is weird. I only get the bug I mentioned with the "terrain height" variable, and I don't know why.
Here is the full Traceback in an Ipython session.
It might be related to the datatype of terrain height. It is likely that that variable is not really needed, but I would rather have it as an option (since a user might want to know the height of the terrain at a certain location according to TEMPO).

```
Cell In[2], line 1
----> 1 vals = mio.sat._tempo_l2_no2_mm.open_dataset(
      2     "/glade/campaign/acom/acom-da/sma/TEMPO_NO2_V03/TEMPO_NO2_L2_V03_20240707*S011G02*.nc",
      3     {
      4         "vertical_column_stratosphere": {"minimum":0},
      5         "scattering_weights": {},
      6         "vertical_column_troposphere": {},
      7         "main_data_quality_flag": {"quality_flag_max": 0},
      8         #"surface_pressure": {"maximum": 100000},
      9         "terrain_height": {"minimum": 0}
     10     },
     11 )

File ~/melodies-monet/monetio/monetio/sat/_tempo_l2_no2_mm.py:230, in open_dataset(fnames, variable_dict, debug)
    227 granules = OrderedDict()
    229 for file in files:
--> 230     granule = _open_one_dataset(file, variable_dict)
    231     apply_quality_flag(granule)
    232     granules[granule.attrs["reference_time_string"]] = granule

File ~/melodies-monet/monetio/monetio/sat/_tempo_l2_no2_mm.py:118, in _open_one_dataset(fname, variable_dict)
    116     minimum = variable_dict[varname]["minimum"]
    117     #values[:] = np.where(values[:] >= minimum, values[:], np.nan)
--> 118     values[values < minimum] = np.nan
    120 if "maximum" in variable_dict[varname]:
    121     maximum = variable_dict[varname]["maximum"]

File /glade/work/plichtig/conda-envs/melodies-monet-test/lib/python3.11/contextlib.py:81, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     78 @wraps(func)
     79 def inner(*args, **kwds):
     80     with self._recreate_cm():
---> 81         return func(*args, **kwds)

File /glade/work/plichtig/conda-envs/melodies-monet-test/lib/python3.11/site-packages/numpy/ma/core.py:3387, in MaskedArray.__setitem__(self, indx, value)
   3384     mval = tuple([False] * len(_dtype.names))
   3385 if _mask is nomask:
   3386     # Set the data, then the mask
-> 3387     _data[indx] = dval
   3388     if mval is not nomask:
   3389         _mask = self._mask = make_mask_none(self.shape, _dtype)

ValueError: cannot convert float NaN to integer
```

EDIT: I con confirm that this is the case:

```
In [4]: np.nan < 3
Out[4]: False

In [5]: a = np.array([1,2,3,4,5], dtype=int)

In [6]: 

In [6]: a<np.nan
Out[6]: array([False, False, False, False, False])

In [7]: a[a<np.nan] = np.nan
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[7], line 1
----> 1 a[a<np.nan] = np.nan

ValueError: cannot convert float NaN to integer
```

This does not happen for floats.
I'll work out a way to deal with this and update this again

zmoon replied:
Weird that terrain height would be int. But yes I see the issue. Easiest solution would be to convert the array to float first (`.astype(float)`) before NaN application. But it looks this may actually be a masked array, so you'd really want to modify the mask, not set values to NaN.

Thanks for working on this, which relates to #184 as well. I encountered a masked int variable [there](https://github.com/noaa-oar-arl/monetio/blob/0a93c4396057257d654d75e44b08ba49d34a1618/monetio/sat/_tropomi_l2_no2_mm.py#L151) too and was just skipping it, but maybe `.astype(float).filled(np.nan)` conversion should be done instead.

blychs replied:
Casting it is definitely a solution, but another (possibly simpler) solution is just inverting the condition of my previous code. You can look at my next commit for that. I. e., it is doing


values = np.where(values [:] < min, np.nan, values[:])

Since the condition is always false if min==np.nan , this would avoid the previous issue, and wouldn't require casting the array into float.

Although, of course, this is still modifying the values instead of the mask. I should probably take your advice and look at how to modify the mask.
terrain_height is of type "short" (ie, float32), so I'm not sure why it fails as if it were int, but it is the exact same error message. Maybe it's a "feature" of masked arrays?

blychs replied:
Done. Now it deals with "maximum" and "minimum" by masking the arrays. I don't know if this really makes a difference at the xarray.DataArray level.

blychs committed:
Debug min/max/quality_flag being NaN, decode time

	Preserve the correct behaviour of the "minimum" and "maximum"
	flags when comparing arrays to those values, while making
	sure it still works with int type arrays (np.NaN can't be
	casted to hace a[a < min] = np.nan if dtype(a) = np.array(int).

	Decode time.

blychs committed:
Make "minimum" and "maximum" flags mask array

	Previous behaviour was switching values into NaN.
	New behaviour is consistent with data types.

blychs committed:
Better handle pressure, convert from hPa to Pa

	Make sure that the code does not force the user to have the
	3D pressure if they did not ask for it.

blychs committed:
format with pre-commit hooks

blychs committed:
add TEMPO L2 NO2 test

blychs committed:
Fix always trying to find surface_pressure

blychs committed:
make sure test catches expected warning

blychs committed:
Remove unnecesary print and warn statements

blychs committed:
remove lat and lon masking

blychs committed:
Remove (wrong) _FillValue guard

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,246 @@
+        elif varname in ["fit_rms_residual", "fit_convergence_flag"]:
+            values_var = dso.groups["qa_statistics"][varname]
+        values = values_var[:].squeeze()
+        fv = values_var.getncattr("_FillValue")
+        if not np.isfinite(fv):
+            values[:][values[:] == fv] = np.nan
```
```suggestion
```
From what I've read, this is done automatically by netcdf4-python (though it uses masked array to do it), so no need to redo it here.

blychs committed:
Manage time whithoug xr.decode_cf()

zmoon commented on tests/test_tempo_l2.py:
```diff
@@ -0,0 +1,80 @@
+        "time",
+        "vertical_column_troposphere",
+    }
+    assert set(ds2["pressure"].dims) == {"swt_level_stagg", "x", "y"}
```
```suggestion
    assert ds2["pressure"].dims == ("swt_level_stagg", "x", "y")
```
tuple to test order too

blychs replied:
Currently they are not in that order, but I'm happy to change that to make it consistent with other readers

zmoon replied:
You could change the test instead, either way

zmoon commented on tests/test_tempo_l2.py:
```diff
@@ -0,0 +1,80 @@
+    assert set(ds) == {vn}
+    assert set(ds.attrs) == {"granule_number", "reference_time_string", "scan_num"}
+
+    with pytest.warns(UserWarning):
```
Can you add a `match="..."` to be more specific about the warning being captured?

blychs committed:
Make more consistent with the TROPOMI reader

	Change np.masked_where to np.masked_less and np.masked_greater.
	Change the order of the dimensions of pressure.

blychs committed:
Update tests/test_tempo_l2.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs committed:
Fix bug in change of dim order for press

blychs committed:
Test also order of dimensions of pressure

zmoon commented on monetio/sat/__init__.py:
```diff
@@ -13,6 +14,7 @@
 )
 
 __all__ = [
+    "_tempo_l2_no2_mm",
```
Could you move this to the alpha order position like in the imports?

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,246 @@
+    ds["time"] = (
+        ("time",),
+        num2pydate(time_var[:].squeeze(), time_units),
+        {"long_name": time_var.long_name, "units": time_var.units},
```
```suggestion
        {"long_name": time_var.long_name},
```
Units attr no longer relevant after decoding

blychs replied:
I think that the units attribute is worth keeping. However, the way it's now fails at save when it tries to encode it, because it is creating a DatetimeIndex with it's own encoding.
I'll find a way to fix it and commit again. Sorry for this.

zmoon replied:
You could call it "original_units" instead?

blychs replied:
The units need to be the real units, that will be found in the attributes of the netcdf that gets created if someone saves it. The problems seems to lie in the fact that xarray is creating and index with time, not just the (expected) coordinates and variable. I need to find out a way to get rid of that.
Normally, with xarray, if you set valid units for time, it should be able to take care of applying the correct encoding on its own.

blychs replied:
OK, you were right and I was wrong about this. XArray manages the encoding on it's own, and even though I can force it (by setting an ds["time"].encoding = {"dtype":  "float32", "units": time_var.units} ) it doesn't seem to make a lot of sense. The only *nicer* thing about doing that is for people to might save de file and want to process it with other tools, that might need consistency in the time units.
Just going with your commit is the right way to go, and I just leave this here in case we need to refer to it in the future.

blychs committed:
Fix order of __all__

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

If set "units" are needed in the future, change ds["time"].encoding

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs committed:
Merge branch 'feature/_tempo_l2_mm_only' of github.com:NCAR/monetio into feature/_tempo_l2_mm_only

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,246 @@
+
+def _open_one_dataset(fname, variable_dict):
+    """Read locally stored INTELSAT TEMPO NO2 level 2
+    Parameters
```
```suggestion

    Parameters
```

zmoon commented on monetio/sat/_tempo_l2_no2_mm.py:
```diff
@@ -0,0 +1,246 @@
+    -------
+    layer_pressure: xr.DataArray
+        Pressure at layer in Pa
+    delta_pressure: xr.DataArray
```
```suggestion
    layer_pressure : xr.DataArray
        Pressure at layer in Pa
    delta_pressure : xr.DataArray
```

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

blychs committed:
Update monetio/sat/_tempo_l2_no2_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon reviewed (approved):
Thanks @blychs for your work on this and helping with #184 as well.
